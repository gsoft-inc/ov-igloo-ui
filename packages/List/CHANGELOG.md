# @igloo-ui/list

## 0.1.1

### Patch Changes

- d5c91f5: link shared components with aliases
- Updated dependencies [d5c91f5]
  - @igloo-ui/checkbox@0.1.8

## 0.1.0

### Minor Changes

- 7dcba93: Initial release of List and Combobox components. Select was updated to use the Dropdown component that was created previously as well as the new List component.
