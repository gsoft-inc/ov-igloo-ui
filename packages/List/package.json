{
  "name": "@igloo-ui/list",
  "version": "0.0.0",
  "main": "dist/List.js",
  "module": "dist/List.js",
  "types": "dist/List.d.ts",
  "styles": "dist/list.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/checkbox": "^0.1.7",
    "@igloo-ui/icons": "^1.6.0",
    "@igloo-ui/tokens": "^2.0.0",
    "classnames": "^2.3.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
