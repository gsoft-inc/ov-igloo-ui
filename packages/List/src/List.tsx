import * as React from "react";
import cx from "classnames";

import ListItem, { type Option, type Member, type OptionType } from "./ListItem";

import "./list.scss";

export interface ListProps extends React.ComponentPropsWithRef<"ul"> {
    /** Add a specific class to the Select */
    className?: string;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** Whether or not to disable tabbing of list items */
    disableTabbing?: boolean;
    /** The option that is currently being focused or hovered */
    focusedOption?: OptionType | null;
    /** True for a compact appearance (In figma, isCompact = "small" and not compact is "medium") */
    isCompact?: boolean;
    /** Whether or not the list is loading */
    loading?: boolean;
    /** The List gains checkboxes beside each option
   * to be able to select multiple options */
    multiple?: boolean;
    /** Called when an option becomes focused or hovered */
    onOptionFocus?: (option: OptionType) => void;
    /** Called when an option is selected */
    onOptionChange?: (option: OptionType) => void;
    /** Called when the mouse moves outside of the option
   * or the option loses focus */
    onOptionBlur?: (option: OptionType) => void;
    /** A list of options */
    options?: OptionType[];
    /** The initial selected option or a list of selected options */
    selectedOption?: OptionType | OptionType[] | null;
    /** Whether or not to show the icon inside the
   * list if it's available */
    showIcon?: boolean;
}

const List: React.FunctionComponent<ListProps> = React.forwardRef(
    (
        {
            className,
            dataTest,
            disableTabbing = false,
            focusedOption,
            isCompact = true,
            loading,
            multiple,
            onOptionFocus,
            onOptionChange,
            onOptionBlur,
            options,
            selectedOption,
            showIcon = true,
            ...rest
        }: ListProps,
        ref: React.ForwardedRef<HTMLUListElement>
    ) => {
        const listClasses = cx("ids-list", className, {
            "ids-list--compact": isCompact,
            "ids-list--multi-select": multiple
        });

        const optionsAreLoading = loading || options?.length === 0;

        const loadingOptions = Array.from({ length: 6 }, (_, index) => (
            <ListItem
                loading
                optionKey={(index + 1).toString()}
                key={(index + 1).toString()}
                isCompact={isCompact}
            />
        ));

        const listItemOptions = options?.map((option: OptionType) => {
            let selected = false;

            if (multiple) {
                if (Array.isArray(selectedOption)) {
                    const selectedItem = selectedOption.filter(o => {
                        return o.value === option.value;
                    });
                    selected = !!selectedItem && selectedItem.length > 0;
                }
            } else if (selectedOption && !Array.isArray(selectedOption)) {
                selected = selectedOption.value === option.value;
            }

            let isFocused = false;
            if (focusedOption) {
                isFocused = focusedOption.value === option.value;
            }

            return (
                <ListItem
                    key={option.value}
                    option={option}
                    onOptionChange={onOptionChange}
                    onOptionFocus={onOptionFocus}
                    onOptionBlur={onOptionBlur}
                    isCompact={isCompact}
                    isFocused={isFocused}
                    isSelected={selected}
                    useCheckbox={multiple}
                    showIcon={showIcon}
                    disableTabbing={disableTabbing}
                />
            );
        });

        return (
            <ul
                ref={ref}
                className={listClasses}
                data-test={dataTest}
                tabIndex={-1}
                role="listbox"
                {...rest}
            >
                {optionsAreLoading ? loadingOptions : listItemOptions}
            </ul>
        );
    }
);

export default List;
export type { Option, Member, OptionType };
