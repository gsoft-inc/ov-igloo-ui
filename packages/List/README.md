# List

Lists render possible actions or selectable options for a given element. This component is usually placed under a kebab menu icon.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/list` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/list
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/list
```

## Usage

Then to use the component in your code just import it!

```jsx
import List from '@igloo-ui/list';

const optionList = [
  {
    type: 'list',
    label: 'Item 1',
    value: '1',
    description: 'just adding some description',
  },
  {
    type: 'list',
    label: 'Item 2 (disabled)',
    value: '2',
    disabled: true,
  },
  {
    type: 'list',
    label: 'Item 3 (focused)',
    value: '3',
    description: 'just adding some description',
  },
  {
    type: 'list',
    label: 'Item 4 (selected)',
    value: '4',
  },
  {
    type: 'list',
    label: 'Item 5',
    value: '5',
    color: '#74DCC9',
  },
];

const [selectedOption, setSeletedOption] = React.useState(optionList[3]);
const [focusedOption, setFocusedOption] = React.useState(optionList[2]);

function handleOptionSelect(option) {
  setSeletedOption(option);
}

function handleOptionFocus(option) {
  setFocusedOption(option);
}

<List
  options={optionList}
  onOptionChange={handleOptionSelect}
  onOptionFocus={handleOptionFocus}
  selectedOption={selectedOption}
  focusedOption={focusedOption}
/>;
```
