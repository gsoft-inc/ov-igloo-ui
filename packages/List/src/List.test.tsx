/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';

import List from './List';

const setup = (props = {}) => {
  return render(
    <List dataTest="ids-list" {...props}>
      Hello world
    </List>
  );
};

describe('List', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-list');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });
});
