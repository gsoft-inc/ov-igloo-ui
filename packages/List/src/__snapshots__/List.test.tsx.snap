// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List It should render a snapshot 1`] = `
<DocumentFragment>
  <ul
    class="ids-list ids-list--compact"
    data-test="ids-list"
    role="listbox"
    tabindex="-1"
  >
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 1
              </span>
              <span
                class="ids-list-item__text-description"
              >
                just adding some description
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact ids-list-item--disabled"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 2 (disabled)
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <div
          class="ids-list-item__visual-identifier-container ids-list-item__svg-container"
        >
          <svg
            class="ids-icon ids-visual-identifier ids-visual-identifier__icon ids-list-item__visual"
            fill="none"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M13.5 8a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0ZM15 8A7 7 0 1 1 1 8a7 7 0 0 1 14 0ZM5.375 6.687a.875.875 0 1 0 0-1.75.875.875 0 0 0 0 1.75Zm6.125-.875a.875.875 0 1 1-1.75 0 .875.875 0 0 1 1.75 0ZM4.805 8.541a.75.75 0 1 0-1.48.248 4.741 4.741 0 0 0 9.35 0 .75.75 0 1 0-1.48-.248 3.241 3.241 0 0 1-6.39 0Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 3 (focused)
              </span>
              <span
                class="ids-list-item__text-description"
              >
                just adding some description
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 4 (selected)
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <div
          class="ids-list-item__visual-identifier-container"
        >
          <div
            class="ids-color ids-color--small ids-visual-identifier ids-visual-identifier__color ids-list-item__visual"
            style="background-color: rgb(116, 220, 201);"
          />
        </div>
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 5
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 6
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
    <li
      aria-selected="false"
      class="ids-list-item ids-list-item--compact"
      role="option"
      tabindex="-1"
    >
      <div
        class="ids-list-item__content"
      >
        <span
          class="ids-list-item__text"
        >
          <span
            class="ids-list-item__text-body"
          >
            <span
              class="ids-list-item__text-content"
            >
              <span
                class="ids-list-item__text-label"
              >
                Item 7
              </span>
            </span>
          </span>
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
